enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1051 @@
+data migration in one direction: forward in regards to Y stream.
+For now, maximum allowed version skew is Y+1, but it might change in the future depending
+on upstream migration rules.
+
+It means that it will be possible to use different Z versions of MicroShift with the same
+data unless there's a breaking change making it impossible, in such case it should be documented.
+
+Given above we can define:
+- rollback as booting older deployment due to admin actions or unhealthy (red) boot
+- upgrade as MicroShift version change from X.Y to X.Y+1 and resulting in data migration
+- downgrade as MicroShift version change from X.Y to X.Y-1 (or older) which is unsupported
+
+Both rollback and downgrade maybe look similar in terms of version change,
+the difference is that for rollback a matching backups exists,
+whereas for downgrade a data migration would have to be performed
+
+Decision to perform or refuse a data migration to schema compatible with newly loaded 
+MicroShift version will be based on following facts:
+- version persisted in MicroShift's data dir (version that created/successfully ran using the data),
+  also referred to as (version) metadata
+- version of currently installed MicroShift binary
+- embedded in MicroShift binary list of blocked "from" versions
+
+A general flow will have following form:
+1. If persisted version is missing, assume 4.13.
+1. If version of `microshift` binary is older than version in metadata, **refuse to start MicroShift**.
+1. If persisted version is on a list of blocked version migrations, **refuse to start MicroShift**.
+1. If binary is the same version as persisted in metadata, **no need for a data migration**.
+1. Otherwise attempt to migrate the data.
+
+
+### Open Questions [optional]
+
+#### Should 4.13 -> 4.14 be supported?
+- Even if we handle "no metadata, no backup, existing data, no next-boot-action" and make
+  a backup, upon rollback 4.13 won't be able to restore it, so it would try to use 4.14's
+  data.
+  - Is manual intervention acceptable (admin manually copying backup to data)?
+  - We most likely don't want to implement special case for that (`pre-run` would save 
+    that persist it was 4.13 previously and red script would restore before shutting down)
+
+#### If green/red info is not persisted, and it's not a first boot (/var/lib/microshift exists), what should we do?
+- Compare versions (metadata vs binary)
+  - Same - start cluster
+  - Different - means that both binaries implement the enhancement,
+    so this could be considered a bug.
+  - Metadata is also missing
+    - Assume data was created by 4.13, do a backup and follow data migration rules
+
+#### When should version metadata file be created
+- By the green script
+  - Existence of file would indicate that data belongs to healthy deployment
+    - What upon reboot? Should file be cleared to be later recreated by green script?
+      - Blind assumption `.version exists == healthy data` might be too implicit and lead
```

Comment:
If the version file does not exist, how do we know if the data is healthy and what version of MicroShift created it?

Reply:
Exactly, that is addressed in second bullet of this open question:
```json
{
  "version": "4.14.0",
  "status": "unknown | healthy | unhealthy"
}
```